modules/library/main/src/main/java/org/geotools/feature/simple/SimpleFeatureTypeBuilder.java

```diff
@@ -1028,7 +1028,10 @@ public static SimpleFeatureType retype(SimpleFeatureType original, List<String>
 
         // add attributes in order
         for (int i = 0; i < types.size(); i++) {
-            b.add(original.getDescriptor(types.get(i)));
+            AttributeDescriptor descriptor = original.getDescriptor(types.get(i));
+            if (descriptor != null) {
+                b.add(descriptor);
+            }
```

Comment:
Okay so this is called by ShapefileFeatureSource.getReadSchema - which already contains logic to deal with converting a blank attribute to the_geom.

Going to fix that workaround rather than make retype perform any logic.